app/js/GcsPersistor.js

```diff
@@ -0,0 +1,291 @@
+      }
+    }
+
+    const uploadStream = storage
+      .bucket(bucket)
+      .file(key)
+      .createWriteStream(writeOptions)
+
+    await pipeline(meteredStream, uploadStream)
+
+    // if we didn't have an md5 hash, we should compare our computed one with Google's
+    // as we couldn't tell GCS about it beforehand
+    if (hashPromise) {
+      sourceMd5 = await hashPromise
+      // throws on mismatch
+      await PersistorHelper.verifyMd5(GcsPersistor, bucket, key, sourceMd5)
+    }
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'upload to GCS failed',
+      { bucket, key },
+      WriteError
+    )
+  }
+}
+
+async function getFileStream(bucket, key, opts = {}) {
+  if (opts.end) {
+    // S3 (and http range headers) treat 'end' as inclusive, so increase this by 1
+    opts.end++
+  }
+  const stream = storage
+    .bucket(bucket)
+    .file(key)
+    .createReadStream(opts)
+
+  const meteredStream = PersistorHelper.getMeteredStream(stream, (_, bytes) => {
+    // ignore the error parameter and just log the byte count
+    metrics.count('gcs.ingress', bytes)
+  })
+
+  try {
+    await PersistorHelper.waitForStreamReady(stream)
+    return meteredStream
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'error reading file from GCS',
+      { bucket, key, opts },
+      ReadError
+    )
+  }
+}
+
+async function getFileSize(bucket, key) {
+  try {
+    const metadata = await storage
+      .bucket(bucket)
+      .file(key)
+      .getMetadata()
+    return metadata[0].size
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'error getting size of GCS object',
+      { bucket, key },
+      ReadError
+    )
+  }
+}
+
+async function getFileMd5Hash(bucket, key) {
+  try {
+    const metadata = await storage
+      .bucket(bucket)
+      .file(key)
+      .getMetadata()
+    return PersistorHelper.base64ToHex(metadata[0].md5Hash)
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'error getting hash of GCS object',
+      { bucket, key },
+      ReadError
+    )
+  }
+}
+
+async function deleteFile(bucket, key) {
+  try {
+    await storage
+      .bucket(bucket)
+      .file(key)
+      .delete()
+  } catch (err) {
+    const error = PersistorHelper.wrapError(
+      err,
+      'error deleting GCS object',
+      { bucket, key },
+      WriteError
+    )
+    if (!(error instanceof NotFoundError)) {
+      throw error
+    }
+  }
+}
+
+async function deleteDirectory(bucket, key) {
+  let files
+
+  try {
+    const response = await storage.bucket(bucket).getFiles({ directory: key })
+    files = response[0]
+  } catch (err) {
+    const error = PersistorHelper.wrapError(
+      err,
+      'failed to list objects in GCS',
+      { bucket, key },
+      ReadError
+    )
+    if (error instanceof NotFoundError) {
+      return
+    }
+    throw error
+  }
+
+  for (const index in files) {
```

Comment:
I think we should use the [Bucket#deleteFiles](https://googleapis.dev/nodejs/storage/latest/Bucket.html#deleteFiles) method provided by Google here (with `force:true`).  It is not atomic but it does allow deleting all files in a bucket under a `prefix/` with a single call. 

It would be a good idea to check that the prefix has the expected form so that we cannot delete the whole bucket by mistake!